File: src/bmt_extract_image.py
import os

def extract_jpeg_from_bmt(bmt_path, output_path):
    """
    Attempts to extract an embedded JPEG from a BMT file by finding the JPEG header.
    """
    try:
        with open(bmt_path, 'rb') as f:
            data = f.read()

        # JPEG Magic Numbers: Start (FF D8), End (FF D9)
        # We look for the first occurrences. Some files might have multiple, 
        # usually the first one is the preview or the main thermal image overlay.
        start_marker = b'\xFF\xD8'
        end_marker = b'\xFF\xD9'

        start_index = data.find(start_marker)
        if start_index == -1:
            print(f"No JPEG start marker found in {bmt_path}")
            return False

        # Search for end marker after start marker
        end_index = data.find(end_marker, start_index)
        
        # JPEG files can contain thumbnails, so we might find an end marker early.
        # However, for a simple extraction, let's grab from start to the last end marker 
        # or try to be smarter if this fails. 
        # Let's search for the LAST end marker to be safe, or just take the first valid block.
        # A safer bet for embedded images is often finding the largest block between markers,
        # but let's try the first one first.
        
        if end_index == -1:
             print(f"No JPEG end marker found in {bmt_path}")
             return False

        jpeg_data = data[start_index:end_index + 2] # Include the end marker

        with open(output_path, 'wb') as out:
            out.write(jpeg_data)
        
        print(f"Successfully extracted image to {output_path}")
        return True

    except Exception as e:
        print(f"Error processing {bmt_path}: {e}")
        return False

if __name__ == "__main__":
    # Test with the file we know exists
    test_file = r"D:\TOOL GOOGLE ANTIGRAVITY\5. Tool Testo\1. Database\Image Site\KHO\1637\Testo Images\IR001366.BMT"
    output_test = "test_extracted.jpg"
    
    if os.path.exists(test_file):
        extract_jpeg_from_bmt(test_file, output_test)
    else:
        print(f"Test file not found: {test_file}")

File: src/bmtExtractImage.js
const fs = require('fs');
const path = require('path');

/**
 * Extract the first JPEG image from a BMT file (thermal image)
 */
function extractJpegFromBmt(bmtPath, outputPath) {
    try {
        const data = fs.readFileSync(bmtPath);

        // JPEG Magic Numbers: Start (FF D8), End (FF D9)
        const startMarker = Buffer.from([0xFF, 0xD8]);
        const endMarker = Buffer.from([0xFF, 0xD9]);

        const startIndex = data.indexOf(startMarker);
        if (startIndex === -1) {
            console.error(`No JPEG start marker found in ${bmtPath}`);
            return false;
        }

        // Search for end marker after startup
        let endIndex = data.indexOf(endMarker, startIndex);

        if (endIndex === -1) {
            console.error(`No JPEG end marker found in ${bmtPath}`);
            return false;
        }

        // Include the end marker (2 bytes)
        const jpegData = data.slice(startIndex, endIndex + 2);

        fs.writeFileSync(outputPath, jpegData);

        console.log(`Successfully extracted thermal image to ${outputPath}`);
        return true;

    } catch (e) {
        console.error(`Error processing ${bmtPath}: ${e.message}`);
        return false;
    }
}

/**
 * Extract ALL JPEG images from a BMT file
 * Returns an object with paths: { thermal: string, real: string | null }
 */
function extractAllJpegsFromBmt(bmtPath, outputDir) {
    try {
        const data = fs.readFileSync(bmtPath);
        const baseName = path.basename(bmtPath, path.extname(bmtPath));

        // JPEG Magic Numbers: Start (FF D8), End (FF D9)
        const startMarker = Buffer.from([0xFF, 0xD8]);
        const endMarker = Buffer.from([0xFF, 0xD9]);

        const results = {
            thermal: null,
            real: null
        };

        let searchStart = 0;
        let imageIndex = 0;

        while (searchStart < data.length) {
            const startIndex = data.indexOf(startMarker, searchStart);
            if (startIndex === -1) break;

            const endIndex = data.indexOf(endMarker, startIndex + 2);
            if (endIndex === -1) break;

            // Extract this JPEG
            const jpegData = data.slice(startIndex, endIndex + 2);

            if (imageIndex === 0) {
                // First image is thermal
                const thermalPath = path.join(outputDir, `${baseName}_thermal.jpg`);
                fs.writeFileSync(thermalPath, jpegData);
                results.thermal = thermalPath;
                console.log(`Extracted thermal image: ${thermalPath}`);
            } else if (imageIndex === 1) {
                // Second image is real/visible light
                const realPath = path.join(outputDir, `${baseName}_real.jpg`);
                fs.writeFileSync(realPath, jpegData);
                results.real = realPath;
                console.log(`Extracted real image: ${realPath}`);
            }

            imageIndex++;
            searchStart = endIndex + 2;

            // Most BMT files have only 2 images, stop after finding both
            if (imageIndex >= 2) break;
        }

        if (imageIndex === 0) {
            console.error(`No JPEG images found in ${bmtPath}`);
        } else {
            console.log(`Found ${imageIndex} image(s) in ${bmtPath}`);
        }

        return results;

    } catch (e) {
        console.error(`Error processing ${bmtPath}: ${e.message}`);
        return { thermal: null, real: null };
    }
}

// Test if run directly
if (require.main === module) {
    const testFile = String.raw`D:\TOOL GOOGLE ANTIGRAVITY\5. Tool Testo\1. Database\Image Site\KHO\1637\Testo Images\IR001366.BMT`;
    const outputDir = ".";

    if (fs.existsSync(testFile)) {
        const results = extractAllJpegsFromBmt(testFile, outputDir);
        console.log("Extraction results:", results);
    } else {
        console.error(`Test file not found: ${testFile}`);
    }
}

module.exports = { extractJpegFromBmt, extractAllJpegsFromBmt };
